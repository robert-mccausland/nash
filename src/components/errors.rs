use std::{error::Error, fmt::Display};

use crate::lexer::Token;

#[derive(Debug, PartialEq, Eq)]
pub struct ParserError {
    pub message: String,
    pub start: Option<usize>,
    pub end: Option<usize>,
}

impl ParserError {
    pub fn new(message: String) -> Self {
        Self {
            message,
            start: None,
            end: None,
        }
    }

    pub fn set_position<'a>(&mut self, token: &'a Token<'a>) {
        self.start = Some(token.start);
        self.end = Some(token.end);
    }
}

impl From<&str> for ParserError {
    fn from(value: &str) -> Self {
        ParserError::new(value.to_owned())
    }
}

impl From<String> for ParserError {
    fn from(value: String) -> Self {
        ParserError::new(value)
    }
}

impl Display for ParserError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.message)
    }
}

impl Error for ParserError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        None
    }

    fn description(&self) -> &str {
        "description() is deprecated; use Display"
    }

    fn cause(&self) -> Option<&dyn Error> {
        self.source()
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct ExecutionError {
    message: String,
}

impl ExecutionError {
    fn new(message: String) -> Self {
        Self { message }
    }
}

impl From<&str> for ExecutionError {
    fn from(value: &str) -> Self {
        ExecutionError::new(value.to_owned())
    }
}

impl From<String> for ExecutionError {
    fn from(value: String) -> Self {
        ExecutionError::new(value)
    }
}

impl Display for ExecutionError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.message)
    }
}

impl Error for ExecutionError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        None
    }

    fn description(&self) -> &str {
        "description() is deprecated; use Display"
    }

    fn cause(&self) -> Option<&dyn Error> {
        self.source()
    }
}
