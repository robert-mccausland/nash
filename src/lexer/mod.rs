use std::iter::Peekable;

use serde::Serialize;
use unicode_segmentation::{GraphemeIndices, UnicodeSegmentation};

use crate::errors::LexerError;

mod token_kinds;

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct Token<'a> {
    pub value: TokenValue<'a>,
    pub start: usize,
    pub end: usize,
}

impl<'a> Token<'a> {
    pub fn new(value: TokenValue<'a>, start: usize, end: usize) -> Self {
        Self { value, start, end }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub enum TokenValue<'a> {
    StringLiteral(&'a str),
    IntegerLiteral(&'a str),
    Identifier(&'a str),
    Keyword(&'a str),
    Equals(),
    Plus(),
    LeftBracket(),
    RightBracket(),
    LeftCurly(),
    RightCurly(),
    LeftAngle(),
    RightAngle(),
    LeftSquare(),
    RightSquare(),
    Question(),
    Dot(),
    Colon(),
    Semicolon(),
    DoubleQuote(),
    Comma(),
    Dollar(),
    Bang(),
    Backtick(),
}

#[derive(Debug, PartialEq, Eq)]
enum LexerContext {
    Root,
    Comment,
    String,
    Command,
    TemplateVariable,
}

pub fn lex<'a>(file: &'a str) -> Tokens<'a> {
    Tokens::new(file)
}

pub struct Tokens<'a> {
    buffer: &'a str,
    next: (usize, &'a str),
    iterator: Peekable<GraphemeIndices<'a>>,
    context_stack: Vec<LexerContext>,
}

impl<'a> Tokens<'a> {
    pub fn new(file: &'a str) -> Self {
        Self {
            buffer: file,
            next: (0, ""),
            iterator: file.grapheme_indices(true).peekable(),
            context_stack: vec![LexerContext::Root],
        }
    }
}

impl<'a> Tokens<'a> {
    fn next(&mut self) -> Option<Result<Token<'a>, LexerError>> {
        loop {
            // If we are at the end of the file then the iterator is finished.
            let Some(next) = self.iterator.peek() else {
                return None;
            };

            self.next = *next;

            // If the advance method returns None it means that we need to call it again
            // as it found a token that should be ignored (e.g. whitespace)
            let next = self.advance();
            if let Ok(Some(token)) = next {
                return Some(Ok(token));
            }

            if let Err(mut err) = next {
                if err.position.is_none() {
                    err.position = Some(self.next.0);
                }
                return Some(Err(err));
            }
        }
    }

    fn advance(&mut self) -> Result<Option<Token<'a>>, LexerError> {
        let start = self.next.0;
        let mut end = start + self.next.1.len();
        let mut result = None;

        loop {
            let value = &self.buffer[start..end];

            // Try to parse the next token, returning the previously parsed token if we can't parse
            // it this time.
            let Some(new_result) = token_kinds::try_get_token_kind(&mut self.context_stack, value)
            else {
                return match result {
                    Some(result) => Ok(result),
                    None => Err("Could not parse token.".into()),
                };
            };

            result = Some(
                new_result
                    .into_token(&value)
                    .map(|value| Token::new(value, start, end)),
            );

            // Advance though the file if we did successfully parse the previous token to see if
            // the next character also makes a valid token.
            self.iterator.next();
            let Some(next) = self.iterator.peek() else {
                return match result {
                    Some(result) => Ok(result),
                    None => Err("Unexpected end of file.".into()),
                };
            };
            self.next = *next;
            end += self.next.1.len();
        }
    }
}

impl<'a> Iterator for Tokens<'a> {
    type Item = Result<Token<'a>, LexerError>;

    fn next(&mut self) -> Option<Self::Item> {
        self.next()
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot;

    use super::*;

    #[test]
    fn should_tokenize_valid_file() {
        let test_file = r#"
# Comments are fun!
func main() {
  var test_identifier = "Blue \"cheese\" and rice!";
  if 1 + 1 == 2 {
    out(test_identifier);
  };

  exec `echo something`;
}
"#;
        let tokens = lex(test_file).collect::<Result<Vec<_>, _>>().unwrap();
        assert_yaml_snapshot!(tokens);
    }

    #[test]
    fn should_parse_empty_string() {
        let test_file = r#"var test = "";"#;
        let tokens = lex(test_file).collect::<Result<Vec<_>, _>>().unwrap();
        assert_yaml_snapshot!(tokens);
    }

    #[test]
    fn should_parse_template_string() {
        let test_file = r#"var test = "hello ${value}!";"#;
        let tokens = lex(test_file).collect::<Result<Vec<_>, _>>().unwrap();
        assert_yaml_snapshot!(tokens);
    }

    #[test]
    fn should_parse_template_string_with_keyword_substring() {
        let test_file = r#""${index}""#;
        let tokens = lex(test_file).collect::<Result<Vec<_>, _>>().unwrap();
        assert_yaml_snapshot!(tokens);
    }
}
