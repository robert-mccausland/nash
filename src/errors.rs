use std::{error::Error, fmt::Display};

use serde::Serialize;

use crate::lexer::Token;

macro_rules! impl_error {
    ($error:ident) => {
        impl Error for $error {
            fn source(&self) -> Option<&(dyn Error + 'static)> {
                None
            }

            fn description(&self) -> &str {
                "description() is deprecated; use Display"
            }

            fn cause(&self) -> Option<&dyn Error> {
                self.source()
            }
        }

        impl From<&str> for $error {
            fn from(value: &str) -> Self {
                value.to_owned().into()
            }
        }

        impl From<String> for $error {
            fn from(value: String) -> Self {
                Self::new(value)
            }
        }
    };
}

macro_rules! nash_error {
    ($($error:ident,)*) => {
        #[derive(Debug, PartialEq, Eq, Serialize)]
        pub enum NashError {
            $(
                $error($error),
            )*
            Other(String)
        }

        impl NashError {
            fn new(value: String) -> Self {
                Self::Other(value)
            }
        }

        impl Display for NashError {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                match self {
                    $(
                        Self::$error(value) => value.fmt(f),
                    )*
                    Self::Other(value) => value.fmt(f),
                }
            }
        }

        $(
            impl From<$error> for NashError {
                fn from(value: $error) -> Self {
                    Self::$error(value)
                }
            }
        )*

        impl_error!(NashError);
    };
}

nash_error![LexerError, ParserError, ExecutionError,];

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct LexerError {
    pub message: String,
    pub position: Option<usize>,
}

impl LexerError {
    pub fn new(message: String) -> Self {
        Self {
            message,
            position: None,
        }
    }
}

impl Display for LexerError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.message.fmt(f)
    }
}

impl_error!(LexerError);

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct ParserError {
    pub message: String,
    pub token: String,
    pub start: Option<usize>,
    pub end: Option<usize>,
}

impl ParserError {
    pub fn new(message: String) -> Self {
        Self {
            message,
            token: String::new(),
            start: None,
            end: None,
        }
    }

    pub fn set_position<'a>(&mut self, token: &'a Token<'a>) {
        self.token = format!("{:?}", token.value);
        self.start = Some(token.start);
        self.end = Some(token.end);
    }
}

impl Display for ParserError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.message.fmt(f)
    }
}

impl_error!(ParserError);

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct ExecutionError {
    message: String,
}

impl ExecutionError {
    fn new(message: String) -> Self {
        Self { message }
    }
}

impl Display for ExecutionError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.message.fmt(f)
    }
}

impl_error!(ExecutionError);
