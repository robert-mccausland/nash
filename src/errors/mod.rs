use std::{error::Error, fmt::Display};

use serde::Serialize;

use crate::lexer::Token;

macro_rules! impl_error {
    ($error:ident $(,$exit_code:expr)?) => {
        impl Error for $error {
            fn source(&self) -> Option<&(dyn Error + 'static)> {
                None
            }

            fn description(&self) -> &str {
                "description() is deprecated; use Display"
            }

            fn cause(&self) -> Option<&dyn Error> {
                self.source()
            }
        }

        impl From<&str> for $error {
            fn from(value: &str) -> Self {
                value.to_owned().into()
            }
        }

        impl From<String> for $error {
            fn from(value: String) -> Self {
                Self::new(value)
            }
        }

        $(
            impl $error {
                pub fn exit_code(&self) -> u8 {
                    $exit_code
                }
            }
        )?
    };
}

macro_rules! nash_error {
    ($($error:ident,)*) => {
        #[derive(Debug, PartialEq, Eq, Serialize)]
        pub enum NashError {
            $(
                $error($error),
            )*
            Other(String)
        }

        impl NashError {
            fn new(value: String) -> Self {
                Self::Other(value)
            }

            pub fn exit_code(&self) -> u8 {
                match self {
                    $(
                        Self::$error(value) => value.exit_code(),
                    )*
                    Self::Other(_) => 100
                }
            }
        }

        impl Display for NashError {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                match self {
                    $(
                        Self::$error(value) => value.fmt(f),
                    )*
                    Self::Other(value) => value.fmt(f),
                }
            }
        }

        $(
            impl From<$error> for NashError {
                fn from(value: $error) -> Self {
                    Self::$error(value)
                }
            }
        )*

        impl_error!(NashError);
    };
}

nash_error![LexerError, ParserError, ExecutionError,];

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct LexerError {
    pub message: String,
    pub position: Option<usize>,
}

impl LexerError {
    pub fn new(message: String) -> Self {
        Self {
            message,
            position: None,
        }
    }
}

impl Display for LexerError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.message.fmt(f)
    }
}

impl_error!(LexerError, 101);

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct ParserError {
    pub message: String,
    pub token: String,
    pub start: Option<usize>,
    pub end: Option<usize>,
}

impl ParserError {
    pub fn new(message: String) -> Self {
        Self {
            message,
            token: String::new(),
            start: None,
            end: None,
        }
    }

    pub fn set_position<'a>(&mut self, token: &'a Token<'a>) {
        self.token = format!("{:?}", token.value);
        self.start = Some(token.start);
        self.end = Some(token.end);
    }
}

impl Display for ParserError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.message.fmt(f)
    }
}

impl_error!(ParserError, 102);

#[derive(Debug, PartialEq, Eq, Serialize)]
pub struct ExecutionError {
    pub message: String,
    pub call_stack: Option<Vec<String>>,
}

impl ExecutionError {
    fn new(message: String) -> Self {
        Self {
            message,
            call_stack: None,
        }
    }

    pub fn set_call_stack(&mut self, call_stack: Vec<String>) {
        self.call_stack = Some(call_stack);
    }
}

impl Display for ExecutionError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.message.fmt(f)
    }
}

impl_error!(ExecutionError, 103);
