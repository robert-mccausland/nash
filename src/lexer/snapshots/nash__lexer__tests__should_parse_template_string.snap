---
source: src/lexer/mod.rs
expression: tokens
---
- value:
    Keyword: var
  start: 0
  end: 3
- value:
    Identifier: test
  start: 4
  end: 8
- value:
    Equals: []
  start: 9
  end: 10
- value:
    DoubleQuote: []
  start: 11
  end: 12
- value:
    StringLiteral: "hello "
  start: 12
  end: 18
- value:
    Dollar: []
  start: 18
  end: 19
- value:
    LeftCurly: []
  start: 19
  end: 20
- value:
    Identifier: value
  start: 20
  end: 25
- value:
    RightCurly: []
  start: 25
  end: 26
- value:
    StringLiteral: "!"
  start: 26
  end: 27
- value:
    DoubleQuote: []
  start: 27
  end: 28
- value:
    Semicolon: []
  start: 28
  end: 29
